sharding/proposer/proposer.go

```diff
@@ -0,0 +1,90 @@
+package proposer
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/node"
+)
+
+// createCollation creates collation base struct with header
+// and body. Header consists of shardID, ChunkRoot, period,
+// proposer addr and signatures. Body contains serialized blob
+// of a collations transactions.
+func createCollation(n node.Node, shardId *big.Int, period *big.Int, txs []*types.Transaction) (*sharding.Collation, error) {
+	// shardId has to be within range
+	if shardId.Cmp(big.NewInt(0)) < 0 || shardId.Cmp(big.NewInt(sharding.ShardCount)) > 0 {
+		return nil, fmt.Errorf("can't create collation for shard %v. Must be between 0 and %v", shardId, sharding.ShardCount)
+	}
+
+	// check with SMC to see if we can add the header.
+	if a, _ := checkHeaderAvailability(n, shardId, period); !a {
+		return nil, fmt.Errorf("can't create collation, collation with same period has already been added")
+	}
+
+	// serialized tx to blob for collation body.
+	blobs, err := sharding.SerializeTxToBlob(txs)
+	if err != nil {
+		return nil, fmt.Errorf("can't create collation, serialization to blob failed: %v", err)
+	}
+
+	// construct the header, leave chunkRoot and signature fields empty, to be filled later.
+	addr := n.Account().Address
+	header := sharding.NewCollationHeader(shardId, nil, period, &addr, nil)
+
+	// construct the body with header, blobs(serialized txs) and txs.
+	collation := sharding.NewCollation(header, blobs, txs)
+	collation.CalculateChunkRoot()
+	sig, err := n.Sign(collation.Header().Hash())
+	if err != nil {
+		return nil, fmt.Errorf("can't create collation, sign collationHeader failed: %v", err)
+	}
+
+	// add proposer signature to collation header.
+	collation.Header().AddSig(sig)
+
+	return collation, nil
+}
+
+// addHeader adds the collation header to the main chain by sending
+// an addHeader transaction to the sharding manager contract.
+// There can only exist one header per period per shard, it's proposer's
+// responsibility to check if a header has been added.
+func addHeader(n node.Node, collation sharding.Collation) error {
+	log.Info("Adding header to SMC")
+
+	txOps, err := n.CreateTXOpts(big.NewInt(0))
+	if err != nil {
+		return fmt.Errorf("unable to initiate add header transaction: %v", err)
+	}
+
+	var chunkRoot [32]byte
+	copy(chunkRoot[:], collation.Header().ChunkRoot().Bytes())
+
+	_, err = n.SMCTransactor().AddHeader(txOps, collation.Header().ShardID(), collation.Header().Period(), chunkRoot)
```

Comment:
Why are you throwing away the transaction? Did it succeed? Has it been mined?

Reply:
I'll add a check to make sure tx succeeds